design/automatic-network-configuration.md

```diff
@@ -0,0 +1,738 @@
+  (optional).
+2. User specifies the `spec.networkConfiguration` field.
+3. User creates `SwitchPortConfiguration` for the deployment.
+4. CAPI / user creates the Metal3Machine (based on the Metal3MachineTemplate),
+  referencing the `SwitchPortConfiguration`.
+5. CAPM3 filters the BMH by calling the `filter()` function. Then select a BMH
+  according to the original method.
+6. Before provisioning, if the machine's `networkConfiguration` field is not
+  empty, CAPM3 will call the `configureNetwork()` method to fill
+  networkConfiguration into the appropriate BMH spec.
+    1. CAPM3 filters the interfaces of BMH based on the NicHint provided
+    for the network configuration.
+    2. CAPM3 identifies the port in the BMH spec that maps to the interface
+    selected based on the NicHint.
+    3. CAPM3 sets the `BMH.spec.ports.configuration` field.
+7. BMO detects that the status.state of the Port CR corresponding to all BMH
+  network cards is empty (without config in the specs) or Active, and then
+  continues to provision BMH.Configure the network configuration set by the
+  user after provision is completed.
+7. BMO edit the SwitchPort.
+    1. BMO fetches the SwitchPort object referred to in the BMH spec.
+    2. BMO sets the SwitchPortConfig in the port specs, and applies it.
+    3. BMO waits for the port to be configured.
+8. The SwitchPort controller detects that SwitchPort.spec.configurationRef changed,
+  then carry out the corresponding processing:
+    1. Verify the user configuration to match the switch configuration
+    (allowed vlans etc.) (Validating step).
+    2. Credentials verification (Validating step).
+    3. Apply the configuration to the port on the switch (Configuring step).
+    4. Update the port CR (status) (Configuring step).
+    5. Port is in Active step.
+
+![](./images/automatic-network-configuration-provisioning.png)
+
+#### Update Workflow
+
+1. The Switch controller detects that Port.spec.configurationRef changed, then
+  carry out the corresponding processing:
+    1. Verify the user configuration to match the switch configuration
+    (allowed vlans etc.) (Validating step).
+    2. Credentials verification (Validating step).
+    3. Apply the configuration to the port on the switch (Configuring step).
+    4. Update the port CR (status) (Configuring step).
+    5. Port is in Active step.
+
+#### Deprovisioning workflow
+
+1. CAPI / User set the deletion timestamp on the `Metal3Machine` CR.
+2. CAPM3 calls the `deconfigureNetwork()` to clears the contents of
+  `BMH.ports.configuration`.
+3. BMO deconfigures the ports.
+    1. `BMH.ports.provisioningConfiguration` is empty
+        1. BMO clears the configuration field in the switchPort CR
+          corresponding to the network card.
+        2. SwitchPort controller starts to deconfigure the network by calling
+          `device.DeConfigurePort()`. Return the port to a `cleaning` state.
```

Comment:
What does the "Idle" state look like after cleaning? Is there some default config for ports that are not explicitly configured?
How do you differentiate between ports that are 'deconfigured' and ports that are simply not managed by the controller (e.g. uplink ports)?

Reply:
>What does the "Idle" state look like after cleaning? 

When performing the clean operation, we will call the `delete_port` function of `network_runner`, which will be processed according to the definition of each manufacturer, mainly to delete the vlan-related configuration.

>  Is there some default config for ports that are not explicitly configured?

No, ports without a specified configuration file will not be configured

> How do you differentiate between ports that are 'deconfigured' and ports that are simply not managed by the controller

We have a `Disabled` field in `Switch.ports` to indicate whether this port is available or controllable.